File: src/node/src/generator/polygon/polygon.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "polygon.h"

#include <cstdio>
#include <cstring>

namespace olive
{

const std::string PolygonGenerator::k_points_input = "points_in";
const std::string PolygonGenerator::k_color_input = "color_in";

#define super GeneratorWithMerge

PolygonGenerator::PolygonGenerator()
{
	add_input(k_points_input, NodeValue::k_bezier, Vector2D(0, 0),
			 InputFlags(k_input_flag_array));

	add_input(k_color_input, NodeValue::k_color,
			 Variant::from_value(Color(1.0, 1.0, 1.0)));

	const int k_middle_x = 135;
	const int k_middle_y = 45;
	const int k_bottom_x = 90;
	const int k_bottom_y = 120;
	const int k_top_y = 135;

	// The Default Pentagon(tm)
	input_array_resize(k_points_input, 5);
	set_split_standard_value_on_track(k_points_input, 0, 0, 0);
	set_split_standard_value_on_track(k_points_input, 1, -k_top_y, 0);
	set_split_standard_value_on_track(k_points_input, 0, k_middle_x, 1);
	set_split_standard_value_on_track(k_points_input, 1, -k_middle_y, 1);
	set_split_standard_value_on_track(k_points_input, 0, k_bottom_x, 2);
	set_split_standard_value_on_track(k_points_input, 1, k_bottom_y, 2);
	set_split_standard_value_on_track(k_points_input, 0, -k_bottom_x, 3);
	set_split_standard_value_on_track(k_points_input, 1, k_bottom_y, 3);
	set_split_standard_value_on_track(k_points_input, 0, -k_middle_x, 4);
	set_split_standard_value_on_track(k_points_input, 1, -k_middle_y, 4);

	// Initiate gizmos
	poly_gizmo_ = new PathGizmo(this);
	add_gizmo(poly_gizmo_);
}

std::string PolygonGenerator::name() const
{
	return "Polygon";
}

std::string PolygonGenerator::id() const
{
	return "org.olivevideoeditor.Olive.polygon";
}

std::vector<Node::CategoryID> PolygonGenerator::category() const
{
	return { k_category_generator };
}

std::string PolygonGenerator::description() const
{
	return "Generate a 2D polygon of any amount of points.";
}

void PolygonGenerator::retranslate()
{
	super::retranslate();

	set_input_name(k_points_input, "Points");
	set_input_name(k_color_input, "Color");
}

ShaderJob PolygonGenerator::get_generate_job(const NodeValueRow &value,
										   const VideoParams &params) const
{
	VideoParams p = params;
	p.set_format(PixelFormat::u8);
	auto job = Texture::job(p, GenerateJob(value));

	// Conversion to RGB
	ShaderJob rgb;
	rgb.set_shader_id("rgb");
	rgb.insert("texture_in", NodeValue(NodeValue::k_texture, job, this));
	rgb.insert("color_in", value.at(k_color_input));

	return rgb;
}

void PolygonGenerator::value(const NodeValueRow &value,
							 const NodeGlobals &globals,
							 NodeValueTable *table) const
{
	push_mergable_job(value,
					Texture::job(globals.vparams(),
								 get_generate_job(value, globals.vparams())),
					table);
}

void PolygonGenerator::generate_frame(FramePtr frame,
									 const GenerateJob &job) const
{
	// Formerly drawn with QPainter into a QImage wrapping the frame buffer
	// (QImage::Format_RGBA8888_Premultiplied, filled Qt::transparent). The
	// rasterizer now lives behind the facade-installed PathFillBackend (see
	// node/geometry.h); the path data model below is unchanged.
	unsigned char *img = reinterpret_cast<unsigned char *>(frame->data());
	const int width = frame->width();
	const int height = frame->height();
	const int linesize = frame->linesize_bytes();

	// Clear to transparent (formerly QImage::fill(Qt::transparent))
	for (int y = 0; y < height; y++) {
		std::memset(img + size_t(linesize) * y, 0, size_t(width) * 4);
	}

	if (PathFillBackend fill = path_fill_backend()) {
		auto points = job.get(k_points_input).to_array();

		PainterPath path = generate_path(points, input_array_size(k_points_input));

		double par = frame->video_params().pixel_aspect_ratio().to_double();
		double scale_x = 1.0 / frame->video_params().divider() / par;
		double scale_y = 1.0 / frame->video_params().divider();
		double translate_x = frame->video_params().width() / 2 * par;
		double translate_y = frame->video_params().height() / 2;

		fill(path, scale_x, scale_y, translate_x, translate_y, img, width,
			 height, linesize);
	} else {
		static bool warned = false;
		if (!warned) {
			fprintf(stderr,
					"PolygonGenerator: no PathFillBackend installed, frame left empty\n");
			warned = true;
		}
	}
}

template <typename T> NodeGizmo *PolygonGenerator::create_appropriate_gizmo()
{
	return new T(this);
}

template <> NodeGizmo *PolygonGenerator::create_appropriate_gizmo<PointGizmo>()
{
	return add_draggable_gizmo<PointGizmo>();
}

template <typename T>
void PolygonGenerator::validate_gizmo_vector_size(std::vector<T *> &vec, int new_sz)
{
	int old_sz = int(vec.size());

	if (old_sz != new_sz) {
		if (old_sz > new_sz) {
			for (int i = new_sz; i < old_sz; i++) {
				delete vec.at(i);
			}
		}

		vec.resize(new_sz);

		if (old_sz < new_sz) {
			for (int i = old_sz; i < new_sz; i++) {
				vec[i] = static_cast<T *>(create_appropriate_gizmo<T>());
			}
		}
	}
}

void PolygonGenerator::update_gizmo_positions(const NodeValueRow &row,
											const NodeGlobals &globals)
{
	Vector2D res;
	if (TexturePtr tex = row.at(k_base_input).to_texture()) {
		res = tex->virtual_resolution();
	} else {
		res = globals.square_resolution();
	}

	Imath::V2d half_res(res.x() / 2, res.y() / 2);

	auto points = row.at(k_points_input).to_array();

	int current_pos_sz = int(gizmo_position_handles_.size());

	validate_gizmo_vector_size(gizmo_position_handles_, int(points.size()));
	validate_gizmo_vector_size(gizmo_bezier_handles_, int(points.size()) * 2);
	validate_gizmo_vector_size(gizmo_bezier_lines_, int(points.size()) * 2);

	for (int i = current_pos_sz; i < int(gizmo_position_handles_.size()); i++) {
		gizmo_position_handles_.at(i)->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_points_input, i), 0));
		gizmo_position_handles_.at(i)->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_points_input, i), 1));

		PointGizmo *bez_gizmo1 = gizmo_bezier_handles_.at(i * 2 + 0);
		bez_gizmo1->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_points_input, i), 2));
		bez_gizmo1->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_points_input, i), 3));
		bez_gizmo1->set_shape(PointGizmo::k_circle);
		bez_gizmo1->set_smaller(true);

		PointGizmo *bez_gizmo2 = gizmo_bezier_handles_.at(i * 2 + 1);
		bez_gizmo2->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_points_input, i), 4));
		bez_gizmo2->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_points_input, i), 5));
		bez_gizmo2->set_shape(PointGizmo::k_circle);
		bez_gizmo2->set_smaller(true);
	}

	int pts_sz = input_array_size(k_points_input);
	if (!points.empty()) {
		for (int i = 0; i < pts_sz; i++) {
			const Bezier &pt = points.at(i).to_bezier();

			Imath::V2d main = pt.to_vec() + half_res;
			Imath::V2d cp1 = main + pt.control_point_1_to_vec();
			Imath::V2d cp2 = main + pt.control_point_2_to_vec();

			gizmo_position_handles_[i]->set_point(PointF(main.x, main.y));

			gizmo_bezier_handles_[i * 2]->set_point(PointF(cp1.x, cp1.y));
			gizmo_bezier_lines_[i * 2]->set_line(
				LineF(PointF(main.x, main.y), PointF(cp1.x, cp1.y)));
			gizmo_bezier_handles_[i * 2 + 1]->set_point(PointF(cp2.x, cp2.y));
			gizmo_bezier_lines_[i * 2 + 1]->set_line(
				LineF(PointF(main.x, main.y), PointF(cp2.x, cp2.y)));
		}
	}

	// The PathGizmo no longer stores a path (drawing data belongs to the app
	// layer since the gizmo wave), so the former set_path() call with the
	// half-resolution-translated generate_path() result is gone; the path
	// itself is still built for generate_frame().
}

ShaderCode PolygonGenerator::get_shader_code(const ShaderRequest &request) const
{
	if (request.id == "rgb") {
		return ShaderCode(
			FileFunctions::read_file_as_string(":/shaders/rgb.frag"));
	} else {
		return super::get_shader_code(request);
	}
}

void PolygonGenerator::gizmo_drag_move(double x, double y, int modifiers)
{
	(void) modifiers;

	DraggableGizmo *gizmo = static_cast<DraggableGizmo *>(current_gizmo());

	if (gizmo == poly_gizmo_) {
		// FIXME: Drag all points
	} else {
		NodeInputDragger &x_drag = gizmo->get_draggers()[0];
		NodeInputDragger &y_drag = gizmo->get_draggers()[1];
		x_drag.drag(x_drag.get_start_value().to_double() + x);
		y_drag.drag(y_drag.get_start_value().to_double() + y);
	}
}

void PolygonGenerator::add_point_to_path(PainterPath *path, const Bezier &before,
									  const Bezier &after)
{
	Imath::V2d a = before.to_vec() + before.control_point_2_to_vec();
	Imath::V2d b = after.to_vec() + after.control_point_1_to_vec();
	Imath::V2d c = after.to_vec();

	path->cubic_to(PointF(a.x, a.y), PointF(b.x, b.y), PointF(c.x, c.y));
}

PainterPath PolygonGenerator::generate_path(const NodeValueArray &points,
											int size)
{
	PainterPath path;

	if (!points.empty()) {
		const Bezier &first_pt = points.at(0).to_bezier();
		Imath::V2d v = first_pt.to_vec();
		path.move_to(PointF(v.x, v.y));

		for (int i = 1; i < size; i++) {
			add_point_to_path(&path, points.at(i - 1).to_bezier(),
						   points.at(i).to_bezier());
		}

		add_point_to_path(&path, points.at(size - 1).to_bezier(), first_pt);
	}

	return path;
}

}

File: src/node/rust/src/nodes/polygon.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Polygon generator (C++ `src/node/src/generator/polygon/polygon.{h,cpp}`,
//! `olive::PolygonGenerator`). Extends the C++ `GeneratorWithMerge`
//! base (see [`super::generatorwithmerge`]).

use crate::factory::NodeMeta;
use crate::node::{Category, NodeBehavior, NodeCore};

/// Points array input id (C++ `k_points_input`). Type: bezier; flags:
/// array. The C++ constructor resizes the array to 5 and sets split
/// standard values to form a default pentagon: `(0, -135)`,
/// `(135, -45)`, `(90, 120)`, `(-90, 120)`, `(-135, -45)`.
pub const POINTS_INPUT: &str = "points_in";

/// Color input id (C++ `k_color_input`). Type: color; default
/// `(1.0, 1.0, 1.0)` (opaque white).
pub const COLOR_INPUT: &str = "color_in";

/// Polygon generator node.
///
/// The C++ members `poly_gizmo_` (PathGizmo), `gizmo_position_handles_`,
/// `gizmo_bezier_handles_` and `gizmo_bezier_lines_` are GUI gizmo
/// pointers with no Rust equivalent here; gizmos are tracked in
/// `NodeCore::gizmos`, so they are omitted.
pub struct PolygonGenerator;

/// Fragment shader for the `"rgb"` shader id (C++ loads
/// `:/shaders/rgb.frag` in `get_shader_code`), recoloring the rasterized
/// polygon mask with the color input. Text copied verbatim from
/// `engine/shaders/rgb.frag`.
const RGB_SHADER_FRAG: &str = r#"// Input texture
uniform sampler2D texture_in;

// Input texture coordinate
in vec2 ove_texcoord;
out vec4 frag_color;

// Input color
uniform vec4 color_in;

void main() {
  vec4 color = texture(texture_in, ove_texcoord);
  color.rgb = color_in.rgb * color.a;
  frag_color = color;
}
"#;

impl PolygonGenerator {
	/// Fragment shader for the `"rgb"` request (C++
	/// `get_shader_code()` `"rgb"` branch).
	fn rgb_shader_frag() -> &'static str {
		RGB_SHADER_FRAG
	}
}

impl NodeBehavior for PolygonGenerator {
	/// Human-readable name (C++ `name()`).
	fn name(&self) -> &str {
		"Polygon"
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.polygon"
	}

	/// Categories (C++ `category()`).
	fn categories(&self) -> &[Category] {
		&[Category::Generator]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"Generate a 2D polygon of any amount of points."
	}

	/// Localized input names (C++ `retranslate()`): the merge-base name
	/// (`base_in` "Base") plus `points_in` -> "Points" and `color_in`
	/// -> "Color".
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		todo!()
	}

	/// Evaluate outputs (C++ `value()`): wraps the generate job
	/// (rasterized at u8 pixel format, then recolored by an `"rgb"`
	/// shader job sampling it as `texture_in` with `color_in`) in a
	/// texture at the sequence video params and pushes it through
	/// `push_mergable_job` (merged over `base_in` when connected).
	fn value(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		time: oakcore_rs::Rational,
		table: &mut crate::value::NodeValueTable,
	) {
		todo!()
	}

	/// Direct frame generation (C++ `generate_frame()`): clears the RGBA
	/// buffer to transparent, builds a closed cubic path through the
	/// bezier points (each segment from the previous point's control
	/// point 2 through the next point's control point 1 to the next
	/// point, closing back to the first), and fills it via the
	/// facade-installed path-fill backend with divider/pixel-aspect
	/// scaling and center translation; without a backend the frame is
	/// left empty (warned once).
	fn generate_frame(
		&self,
		core: &NodeCore,
		frame: &mut crate::bridge::render::TextureHandle,
		time: oakcore_rs::Rational,
	) {
		todo!()
	}

	/// Gizmo layout (C++ `update_gizmo_positions()`): resolution comes
	/// from the base texture's virtual resolution or the square
	/// resolution; grows/shrinks the per-point position handles and the
	/// 2-per-point bezier handles/lines to the point count, wires new
	/// handles to the corresponding `points_in` tracks (position:
	/// tracks 0/1; bezier: tracks 2/3 and 4/5, circle-shaped and
	/// smaller), and places every handle/line at point + half
	/// resolution.
	fn gizmo_update(&self, core: &NodeCore, row: &crate::value::NodeValueRow) {
		todo!()
	}

	/// Gizmo drag (C++ `gizmo_drag_move()`): dragging the path gizmo
	/// itself is a no-op (C++ FIXME: drag all points); dragging any
	/// other gizmo offsets its x/y input draggers by the drag delta
	/// from their start values.
	fn gizmo_drag(&mut self, core: &mut NodeCore, start: bool, x: f64, y: f64, modifiers: u32) {
		todo!()
	}

	/// Shader code request (C++ `get_shader_code()`): `"rgb"` returns
	/// the recolor fragment shader; any other request falls through to
	/// the merge base (`"mrg"` -> alpha-over shader, else empty).
	fn shader_code(&self, request: &str) -> Option<String> {
		todo!()
	}

	/// Deep copy (C++ `copy()`).
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		todo!()
	}
}

/// Constructor (C++ `PolygonGenerator::PolygonGenerator()`): on top of
/// the `GeneratorWithMerge` constructor (which adds `base_in`), adds
/// the `points_in` bezier array and `color_in`, resizes the array to
/// the default pentagon documented on [`POINTS_INPUT`], and creates the
/// path gizmo.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ factory entry for
/// `org.olivevideoeditor.Olive.polygon`).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.polygon",
		name: "Polygon",
		categories: &[Category::Generator],
		create,
	});
}
